File: apps/api/src/cash/expected-cash.ts
export type ShiftCashInputs = {
  openingCash: number;
  saleCashTotal: number;
  saleTransferTotal: number;
  /**
   * SaleReturn.cashRefundVnd cộng dồn trong ca — tiền mặt đã ra khỏi ngăn
   * kéo khi hoàn trả hàng, phải TRỪ khỏi expectedCashVnd. Chỉ phần hoàn
   * bằng tiền mặt; SaleReturn.transferRefundVnd KHÔNG chạm tiền mặt nên
   * không được cộng vào đây (xem loadShiftCashInputsWithClient).
   */
  saleReturnCashTotal: number;
  debtPaymentCashTotal: number;
  debtPaymentTransferTotal: number;
  voucherCashInTotal: number;
  voucherCashOutTotal: number;
  voucherTransferInTotal: number;
  voucherTransferOutTotal: number;
};

export type ShiftCashSnapshot = {
  expectedCashVnd: number;
  transferInShiftVnd: number;
  varianceVnd: number; // closingCash - expected
};

export function computeShiftCashSnapshot(
  inputs: ShiftCashInputs,
  closingCash: number,
): ShiftCashSnapshot {
  const expectedCashVnd =
    inputs.openingCash +
    inputs.saleCashTotal +
    inputs.debtPaymentCashTotal +
    inputs.voucherCashInTotal -
    inputs.voucherCashOutTotal -
    inputs.saleReturnCashTotal;
  const transferInShiftVnd =
    inputs.saleTransferTotal +
    inputs.debtPaymentTransferTotal +
    inputs.voucherTransferInTotal -
    inputs.voucherTransferOutTotal;
  return {
    expectedCashVnd,
    transferInShiftVnd,
    varianceVnd: closingCash - expectedCashVnd,
  };
}

File: apps/pos_app/lib/features/cash/expected_cash.dart
class ShiftCashBreakdown {
  const ShiftCashBreakdown({
    required this.openingCash,
    required this.saleCash,
    required this.debtPaymentCash,
    required this.voucherCashIn,
    required this.voucherCashOut,
    required this.expectedCash,
    required this.saleTransfer,
    required this.debtPaymentTransfer,
    required this.voucherTransferIn,
    required this.voucherTransferOut,
    required this.transferInShift,
  });

  final int openingCash;
  final int saleCash;
  final int debtPaymentCash;
  final int voucherCashIn;
  final int voucherCashOut;
  final int expectedCash;
  final int saleTransfer;
  final int debtPaymentTransfer;
  final int voucherTransferIn;
  final int voucherTransferOut;
  final int transferInShift;

  int variance(int closingCash) => closingCash - expectedCash;
}

ShiftCashBreakdown computeBreakdown({
  required int openingCash,
  required int saleCashTotal,
  required int saleTransferTotal,
  required int debtPaymentCashTotal,
  required int debtPaymentTransferTotal,
  required int voucherCashInTotal,
  required int voucherCashOutTotal,
  required int voucherTransferInTotal,
  required int voucherTransferOutTotal,
}) {
  final expectedCash =
      openingCash +
      saleCashTotal +
      debtPaymentCashTotal +
      voucherCashInTotal -
      voucherCashOutTotal;
  final transferInShift =
      saleTransferTotal +
      debtPaymentTransferTotal +
      voucherTransferInTotal -
      voucherTransferOutTotal;
  return ShiftCashBreakdown(
    openingCash: openingCash,
    saleCash: saleCashTotal,
    debtPaymentCash: debtPaymentCashTotal,
    voucherCashIn: voucherCashInTotal,
    voucherCashOut: voucherCashOutTotal,
    expectedCash: expectedCash,
    saleTransfer: saleTransferTotal,
    debtPaymentTransfer: debtPaymentTransferTotal,
    voucherTransferIn: voucherTransferInTotal,
    voucherTransferOut: voucherTransferOutTotal,
    transferInShift: transferInShift,
  );
}
